In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [5]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
path =  Path("Resources/lending_data.csv")
df_credit = pd.read_csv(path)
# Review the DataFrame
df_credit.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [6]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_credit["loan_status"]

# Separate the X variable, the features
X = df_credit.drop(columns="loan_status")

In [7]:
# Review the y variable Series
y[:10]

0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    0
8    0
9    0
Name: loan_status, dtype: int64

In [8]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [9]:
# Check the balance of our target values
y.value_counts()

loan_status
0    75036
1     2500
Name: count, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [10]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [11]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
logreg_model = logistic_regression_model.fit(X_train, y_train)
logreg_model

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [12]:
# Make a prediction using the testing data
testing_predictions = logreg_model.predict(X_test)
testing_predictions

array([0, 0, 0, ..., 0, 0, 0])

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [16]:
print('Test Accuracy Score: %.3f' % logreg_model.score(X_test, y_test))

Test Accuracy Score: 0.992


In [15]:
# Print the balanced_accuracy score of the model
print('Balanced Accuracy Score: %.3f' % balanced_accuracy_score(y_test, testing_predictions))

Balanced Accuracy Score: 0.952


In [17]:
# Generate a confusion matrix for the model
training_matrix = confusion_matrix(y_test, testing_predictions)
print(training_matrix)

[[18663   102]
 [   56   563]]


In [19]:
# Print the classification report for the model
testing_report = classification_report(y_test, testing_predictions)
print(testing_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** There is a precision of 100% for predicting healthy loans and 85% precision on predicting high risk loans. This means that of all the healthy loans were correctly predicted as healthy, while 15% of the high risk loans were actually healthy loans classified incorrectly. Having such a low number of high risk loans (619) could affect the precision of the model. As more data is gathered, the model will improve its classification of high risk loans.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [13]:
# ************************************************************************************************
# This is omitted per Brett Payne
# ************************************************************************************************
# Per Joanna,
# Brett mentioned in the instructor/TA channels the
# "Predict a Logistic Regression Model with Resampled Training Data" 
# section in the notebook is not required. 
# It used to be taught in this class but isn't anymore.

ModuleNotFoundError: No module named 'imblearn'

In [ ]:
# Count the distinct values of the resampled labels data
# YOUR CODE HERE!

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [ ]:
# ************************************************************************************************
# This is omitted per Brett Payne
# ************************************************************************************************
# Per Joanna,
# Brett mentioned in the instructor/TA channels the
# "Predict a Logistic Regression Model with Resampled Training Data" 
# section in the notebook is not required. 
# It used to be taught in this class but isn't anymore.

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [ ]:
# ************************************************************************************************
# This is omitted per Brett Payne
# ************************************************************************************************
# Per Joanna,
# Brett mentioned in the instructor/TA channels the
# "Predict a Logistic Regression Model with Resampled Training Data" 
# section in the notebook is not required. 
# It used to be taught in this class but isn't anymore.

### Step 4: Answer the following question

# ************************************************************************************************
# This is omitted per Brett Payne
# ************************************************************************************************
# Per Joanna,
# Brett mentioned in the instructor/TA channels the
# "Predict a Logistic Regression Model with Resampled Training Data" 
# section in the notebook is not required. 
# It used to be taught in this class but isn't anymore.**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** YOUR ANSWER HERE!